### PR TITLE
fix/benchmarks

### PR DESCRIPTION
**Pull request by mds1**
Cleans up the elliptic vs. noble benchmarks to ensure no funny business was going on that may have been skewing results, since I would have thought noble is faster all around. Looks like:

- noble is faster for privKey -> pubKey
- elliptic is faster for priveKey * privKey
- elliptic is faster for pubKey + pubKey

Maybe worth opening an issue on the noble repo to discuss and see if there's optimizations that can be made? I trust noble more from a security perspective, but the big difference in multiplication speed is disappointing.

Results from 10k runs are below, in seconds:

|                   | elliptic | noble-secp256k1 |
|-------------------|----------|-----------------|
| privKey -> pubKey | 6.758    | 3.527           |
| privKey * pubKey  | 14.403   | 35.039          |
| pubKey + pubKey   | 0.909    | 1.146           |

### CODE REVIEW

File: benchmark_ec.js [hunk 1]
```diff
@@ -1,86 +1,108 @@
...
-  console.log(`1000x pubKey + pubKey: ${deltaTime} milliseconds`);
+endTime = performance.now()
+deltaTime = endTime - startTime;
+console.log(`1000x privKey --> pubKey: ${deltaTime} milliseconds`);
+
+// --- Private Key * Public Key ---
+startTime = performance.now()
+for (let k = 0; k < numRuns; k++) {
+  const P_scan = secp.getSharedSecret(noblePrivateKeys[k], P_noble);
+  // const P_scan = P_noble.multiply(noblePrivateKeys[k]);
```
Comment (mds1) [inline benchmark_ec.js:95]: Performance is similar for either of these approaches

**Comment by mds1** [comment]
Instead of opening an issue, I'll try summoning @paulmillr here to see if he has any insight 😅

**Comment by paulmillr** [comment]
Elliptic is unsafe. You can easily shut off all precautions in noble and it would probably be faster than elliptic. 
What for?
Do you actually **need** this speed?

**Comment by paulmillr** [comment]
const cached = utils.precompute(pubkey);
cached.multiply(privkey); 

Would also be very fast, and safe.

**Comment by mds1** [comment]
> Do you actually **need** this speed?

The context here is that for stealth address (like [Umbra](https://app.umbra.cash/), here we're working on an EIP to standardize them), the UX bottle neck is scan time: every send emits an event containing a stealth address and ephemeral public key, and you need to scan _all_ events to see if you can compute the private key for that stealth address from your own private key and given public key. The slowest operation during this process is elliptic curve multiplication.

There's various ways to speed this process up, but was surprised to see elliptic's multiplication was faster. For safety reasons I probably wouldn't switch, more interested in seeing if there was a way to safely get comparable performance

> const cached = utils.precompute(pubkey);
> cached.multiply(privkey);
> 
> Would also be very fast, and safe.

This made me realize I need to fix these benchmarks: to more closely mimic our use case we should be holding private key constant and looping over public keys, but currently doing the opposite. Given that, would precomputing still help if the public key changes with each iteration? (planning to test this out, likely won't have time today though)

**Comment by paulmillr** [comment]
> every send emits an event containing a stealth address and ephemeral public key, and you need to scan all events to see if you can compute the private key for that stealth address from your own private key and given public key. The slowest operation during this process is elliptic curve multiplication.

And how many scans need to be executed? Is that 10s, 100s, 1000s, more?

**Comment by mds1** [comment]
Currently for Umbra the worst case is on polygon, where you need to scan ~60k event logs, since that's how many sends there have been, and of course that will only grow over time (for repeat users you can cache to shrink search space, etc.)

**Comment by paulmillr** [comment]
Wow, that's a lot.
Is this the same issue zcash/torn have aka set intersection? 
Have you considered using zcash style diversifiers? (If it's possible at all)

**Comment by mds1** [comment]
Good questions, not sure offhand but will have to look into those

**Comment by nerolation** [comment]
Thanks Matt and Paul. Great information. I agree, noble is then the way to go for further simulations.

**Comment by paulmillr** [comment]
- ~60k event logs means it would take 3.5 min for noble, and 1.5 min for elliptic.
- 600k logs would increase that to 35/15 min: both are unusable from my perspective.
- 6M logs: 5.8/2.5h

See if the `precompute` works for your case. I think it matches your logic?

If not, libsecp256k1 compiled to wasm could work for the first 1M events, after that some other solution would need to be found.

**Comment by mds1** [comment]
@nerolation pushed a commit to delete the other benchmarks file since I don't think we need it anymore

> See if the precompute works for your case. I think it matches your logic?
> 
> If not, libsecp256k1 compiled to wasm could work for the first 1M events, after that some other solution would need to be found.

@paulmillr Planning to give the preocmputing a shot tomorrow, thanks for the help here. I've though about compiling libsecp256k1 to wasm and seeing how that compares, just haven't gotten around to it yet since noble has been sufficient so far

**Pull request merged**